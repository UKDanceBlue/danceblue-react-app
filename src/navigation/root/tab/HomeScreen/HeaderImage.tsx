import { Box, Image, ZStack } from "native-base";

import dbLogo from "../../../../../assets/home/DB_Primary_Logo-01.png";
import backgroundImg from "../../../../../assets/home/db20_ribbon.jpg";

/**
 * A header image container used on the home screen
 */
const HeaderImage = () => (
  <ZStack height="1/2" width="full" m={0}>
    <Image
      source={backgroundImg}
      size="full"
      resizeMode="cover"
      blurRadius={10}
      alt="Picture of DB Marathon" />
    <Box style={{ backgroundColor: "#ffffff55" }}
      width="full"
      height="full" />
    <Image source={dbLogo}
      size="full"
      resizeMode="contain"
      alt="DanceBlue Logo" />
  </ZStack>
);

export default HeaderImage;
