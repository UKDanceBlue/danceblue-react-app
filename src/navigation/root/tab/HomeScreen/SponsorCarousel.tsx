import firebaseFirestore from "@react-native-firebase/firestore";
import { Box, Heading, ScrollView, View } from "native-base";
import { useEffect, useState } from "react";

import SponsorCard from "../../../../common/components/ImageCard";
import { FirestoreSponsor } from "../../../../types/FirebaseTypes";

interface SponsorType extends FirestoreSponsor {
  id: string;
}

/**
 * A horizontally scrolling carousel of SponsorCards
 */
const SponsorCarousel = () => {
  const [ sponsors, setSponsors ] = useState<SponsorType[]>([]);

  useEffect(() => {
    let shouldUpdateState = true;
    async function getSnapshot() {
      const dbSponsors: SponsorType[] = [];
      const snapshot = await firebaseFirestore().collection("sponsors").get();
      snapshot.forEach((document) => {
        dbSponsors.push({ ...document.data(), id: document.id });
      });
      if (shouldUpdateState) {
        setSponsors(dbSponsors);
      }
    }
    void getSnapshot();
    return () => {
      shouldUpdateState = false;
    };
  }, []);

  const cards = sponsors.map((sponsor) => (
    <SponsorCard
      name={sponsor.name ?? ""}
      imagePath={sponsor.logo ?? ""}
      sponsorLink={sponsor.link}
      key={sponsor.id}
    />
  ));

  return (
    <View height="100%" flex={1} alignContent="flex-end">
      <View>
        <Heading> SPONSORS </Heading>
      </View>
      <Box flexDirection="row" flex={1} alignItems="center">
        <ScrollView scrollEventThrottle={16} horizontal flex={1} height="4/5" p="2">
          {cards}
        </ScrollView>
      </Box>
    </View>
  );
};

export default SponsorCarousel;
