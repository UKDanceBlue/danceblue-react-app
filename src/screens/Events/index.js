// Import third-party dependencies
import React from 'react'
import { StyleSheet, ScrollView, SafeAreaView, TouchableOpacity, Text } from 'react-native'
import EventRow from '../../components/event/EventRow'
import { withFirebaseHOC } from '../../../config/Firebase'

// Component for events screen in main navigation
class EventsScreen extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      events: [],
      today: [],
      upcoming: [],
      isLoading: true
    }

    this.splitEventArray = this.splitEventArray.bind(this)
  }

  componentDidMount () {
    const events = []
    this.props.firebase.getUpcomingEvents().then(snapshot => {
      snapshot.forEach(doc =>
        events.push({ id: doc.id, ...doc.data() })
      )
      this.setState({ events: events, isLoading: false }, () => this.splitEventArray())
    })
  }

  splitEventArray () {
    const today = []
    const upcoming = []
    const now = new Date()
    this.state.events.forEach(event => {
      const startTime = event.startTime.toDate()
      if (startTime <= now) today.push(event)
      else upcoming.push(event)
    })
    this.setState({ today: today, upcoming: upcoming })
  }

  render () {
    const { navigate } = this.props.navigation
    return (
      <ScrollView style={styles.body}>
        <SafeAreaView>
          <Text style={styles.heading}>Today's Events</Text>
          {
            this.state.today.map((row) => (
              <TouchableOpacity style={styles.eventRow} onPress={() => navigate('Event', { id: row.id, name: row.title })} key={row.id}>
                <EventRow styles={styles} key={row.id} id={row.id} title={row.title} startDate={row.startTime.toDate()} endDate={row.endTime.toDate()} text={row.text} showIfToday imageLink={row.image} />
              </TouchableOpacity>
            ))
          }
          <Text style={styles.heading}>Upcoming Events</Text>
          {/* jscpd:ignore-start */
            this.state.upcoming.map((row) => (
              <TouchableOpacity style={styles.eventRow} onPress={() => navigate('Event', { id: row.id, name: row.title })} key={row.id}>
                <EventRow styles={styles} key={row.id} id={row.id} title={row.title} startDate={row.startTime.toDate()} endDate={row.endTime.toDate()} text={row.text} showIfToday imageLink={row.image} />
              </TouchableOpacity>
            ))
          /* jscpd:ignore-end */
          }
        </SafeAreaView>
      </ScrollView>
    )
  }
}

EventsScreen.navigationOptions = {
  title: 'Events'
}

const styles = StyleSheet.create({
  eventRow: {
    marginTop: 5,
    marginBottom: 5
  },
  body: {
    padding: 10,
    backgroundColor: 'white',
    flex: 1
  },
  heading: {
    fontSize: 24,
    fontWeight: 'bold'
  }
})

export default withFirebaseHOC(EventsScreen)
