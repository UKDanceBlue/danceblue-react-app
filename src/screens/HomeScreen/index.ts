// Allow importing from the parent folder rather than individual files
// for example you can do:   *import { EventScreen } from '../EventScreen'*
// instead of:               *import { EventScreen } from '../EventScreen/EventScreen'*

import HeaderImage from "./HeaderImage";
import HomeScreen from "./HomeScreen";

export { HeaderImage };
export default HomeScreen;
