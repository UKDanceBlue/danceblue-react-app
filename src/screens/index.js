import React from 'react'
import { StyleSheet, Dimensions, Image } from 'react-native'
import { createStackNavigator } from '@react-navigation/stack'
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs'
import Icon from 'react-native-vector-icons/FontAwesome5'
import { BlurView } from 'expo-blur'

import { withFirebaseHOC } from '../../config/Firebase'

import HomeScreen from './Home'
import {ScoreboardScreen } from './Home'
import EventsScreen from './Events'
import { StoreScreen } from './Store'
import MoreScreen from './More'
import Event from '../components/event/Event'

// import ProfileScreen from './More/Profile'
import { FAQScreen } from './More/FAQ'
import { DonateScreen } from './More/Donate'
import { AboutScreen } from './More/About'
import ProfileScreen from './More/Profile'

import { SplashLoginScreen } from './Modals'

const RootStack = createStackNavigator()
const MainStack = createStackNavigator()
const Tabs = createBottomTabNavigator()

function TabsScreen () {
  return (
    <Tabs.Navigator
      screenOptions={
        ({ route }) => ({
        tabBarIcon: ({ color, size }) => {
          const iconMap = new Map([ //Key: Screen   Value: Icon ID
            ['Home'], ['home'],
            ['Events'], ['calendar'],
            ['Store'], ['store'],
            ['More'], ['ellipsis-h'],
            ['Test'], ['ellipsis-h']
          ]);

          // You can return any component that you like here!
          return (
            <Icon
              name={iconMap.get(route.name)}
              size={size}
              color={color}
              style={{ textAlignVertical: 'center' }}
            />
          )
        }
      })}
      tabBarOptions={{
        activeTintColor: 'white',
        activeBackgroundColor: '#3248a8'
      }}
    >
      <Tabs.Screen name='Home' component={HomeScreen} />
      <Tabs.Screen name='Events' component={EventsScreen} />
      <Tabs.Screen name='Store' component={StoreScreen} />
      <Tabs.Screen name='More' component={MoreScreen} />
    </Tabs.Navigator>
  )
}

function MainStackScreen () {
  return (
    <MainStack.Navigator>
      <MainStack.Screen name='Tab' component={TabsScreen} options={{ headerShown: false }} />
      <MainStack.Screen name='Profile' component={ProfileScreen} />
      <MainStack.Screen name='FAQ' component={FAQScreen} />
      <MainStack.Screen name='Donate' component={DonateScreen} />
      <MainStack.Screen name='About' component={AboutScreen} />
      <MainStack.Screen name='Scoreboard' component={ScoreboardScreen} />
      <MainStack.Screen
        name='Event' component={Event} options={({ route }) => ({
          title: route.params.name,
          headerTransparent: true,
          headerMode: 'screen',
          headerBackground: () => (
            <BlurView tint='light' intensity={100} style={StyleSheet.absoluteFill} />
          )
        })}
      />
    </MainStack.Navigator>
  )
}

class RootStackScreen extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      userID: null,
      isLoading: true
    }
  }

  componentDidMount() {
    this.props.firebase.checkAuthUser(user => {
      if (user !== null) this.setState({ userID: user.uid, isLoading: false })
      else this.setState({ isLoading: false })
    })
  }

  render() {
    if (this.state.isLoading) return <Image style={{ width: Dimensions.get('window').width, height: Dimensions.get('window').height }} source={require('../../assets/splash2.png')} />
    return (
      <RootStack.Navigator mode='modal'>
        {this.state.userID && (
          <RootStack.Screen name='Main' component={MainStackScreen} options={{ headerShown: false }} />
        )}
        {!this.state.userID && (
          <RootStack.Screen name='SplashLogin' component={SplashLoginScreen} options={{ headerShown: false }} />
        )}
      </RootStack.Navigator>
    )
  }
}

export default withFirebaseHOC(RootStackScreen)
