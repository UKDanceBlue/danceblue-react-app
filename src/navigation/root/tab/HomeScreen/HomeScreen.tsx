import { Box, Button, Text, VStack, useTheme } from "native-base";
import { Linking, StatusBar } from "react-native";

import PodcastPlayer from "../../../../common/components/PodcastPlayer";
import SponsorsBlock from "../../../../common/components/SponsorsBlock";
import { useColorModeValue } from "../../../../common/customHooks";
import { universalCatch } from "../../../../common/logging";

import HeaderImage from "./HeaderImage";
import SponsorCarousel from "./SponsorCarousel";

/**
 * Component for home screen in main navigation
 */
const HomeScreen = () => {
  const { colors } = useTheme();
  const bgColor = useColorModeValue(colors.white, colors.gray[600]);

  return (
    <>
      <StatusBar hidden = { false } />
      <VStack flexDirection="column" bgColor={bgColor}>
        <Box height="35%" tintColor={bgColor}>
          <HeaderImage />
        </Box>
        <Box height="10%">
          <Button
            borderRadius={0}
            margin={0}
            backgroundColor={colors.blue[700]}
            _pressed={{ opacity: 0.5 }}
            onPress={async () => {
              if (
                await Linking.canOpenURL(
                  "https://danceblue.networkforgood.com/"
                ).catch(universalCatch)
              ) {
                Linking.openURL(
                  "https://danceblue.networkforgood.com/"
                ).catch(universalCatch);
              }
            }}
          >
            <Text
              bold
              fontSize={30}
              color="light.100"
              shadow="1">Donate Now!</Text>
          </Button>
        </Box>
        <Box height="15%">
          <PodcastPlayer />
        </Box>
        <Box height="40%">
          <SponsorCarousel />
        </Box>
      </VStack>
    </>
  );
};

HomeScreen.navigationOptions = { title: "Home" };

export default HomeScreen;
