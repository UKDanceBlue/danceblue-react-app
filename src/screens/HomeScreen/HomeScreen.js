// Import third-party dependencies
import React, { useEffect, useState } from 'react';
import { SafeAreaView, ScrollView } from 'react-native';

// Import first-party dependencies
import SponsorCarousel from './SponsorCarousel';
import CountdownView from '../../common/components/CountdownView';
import HeaderImage from './HeaderImage';
import { withFirebaseHOC } from '../../firebase/FirebaseContext';

/**
 * Component for home screen in main navigation
 * @param {Object} props Properties of the component: navigation, firebase
 */
const HomeScreen = ({ navigation, navigation: { navigate }, firestore, auth }) => {
  [activeCountdown, setActiveCountDown] = useState(true);
  [userID, setUserID] = useState('');

  useEffect(() => {
    firestore.getConfig().then((doc) => {
      setActiveCountDown(doc.data().activeCountdown);
    });
    auth.checkAuthUser((user) => {
      if (user !== null) {
        if (!user.isAnonymous) {
          setUserID(user.uid);
        }
      }
    });
  });

  return (
    <ScrollView showsVerticalScrollIndicator={false}>
      <SafeAreaView style={{ flex: 1 }}>
        <HeaderImage />
        {activeCountdown && <CountdownView />}
        <SponsorCarousel />
      </SafeAreaView>
    </ScrollView>
  );
};

HomeScreen.navigationOptions = {
  title: 'Home',
};

export default withFirebaseHOC(HomeScreen);
