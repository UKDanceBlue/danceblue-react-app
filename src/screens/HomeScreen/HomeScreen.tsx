import { Linking, SafeAreaView } from "react-native";
import { Button } from "react-native-elements";

import HeaderImage from "./HeaderImage";
import SponsorCarousel from "./SponsorCarousel";

/**
 * Component for home screen in main navigation
 */
const HomeScreen = () => (
  <SafeAreaView style={{ flex: 1 }}>
    <HeaderImage />
    <Button
      style={{ margin: 10, alignSelf: "center" }}
      onPress={() => {
        void (async () => {
          if (
            await Linking.canOpenURL(
              "https://danceblue.networkforgood.com/causes/4789-danceblue-golden-matrix-fund-dance-teams?utm_source=website&utm_medium=unit_website"
            )
          ) {
            void Linking.openURL(
              "https://danceblue.networkforgood.com/causes/4789-danceblue-golden-matrix-fund-dance-teams?utm_source=website&utm_medium=unit_website"
            );
          }
        })();
      }}
      title="Donate!"
    />
    <SponsorCarousel />
  </SafeAreaView>
);

HomeScreen.navigationOptions = { title: "Home" };

export default HomeScreen;
