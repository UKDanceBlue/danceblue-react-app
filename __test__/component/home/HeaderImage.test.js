import React from 'react';
import renderer from 'react-test-renderer';
import { loadAsync } from 'expo-font';
import { FontAwesome5 } from '@expo/vector-icons';
import { HeaderImage } from '../../../src/screens/HomeScreen';

beforeAll(async () => loadAsync(FontAwesome5.font));

test('renders correctly', () => {
  const tree = renderer.create(<HeaderImage />).toJSON();
  expect(tree).toMatchSnapshot();
});
