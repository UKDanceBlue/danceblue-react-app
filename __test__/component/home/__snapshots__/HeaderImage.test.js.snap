// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
      Object {
        "alignSelf": "center",
        "height": 280,
        "overflow": "hidden",
        "width": "100%",
      },
    ]
  }
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      Object {
        "flex": 3,
        "height": "100%",
        "resizeMode": "contain",
        "width": "100%",
      }
    }
  >
    <Image
      source={1}
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": "100%",
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <Image
      source={1}
      style={
        Array [
          Object {
            "flex": 1,
            "height": null,
            "resizeMode": "contain",
            "width": null,
          },
          Object {
            "backgroundColor": "#FFFFFF99",
          },
        ]
      }
    />
  </View>
</View>
`;
